functions/src/judge.ts

```diff
@@ -33,109 +34,191 @@ export async function judge_is_online(_req: Request, res: Response) {
   }
 }
 
+/* eslint-disable  @typescript-eslint/no-explicit-any */
+async function get_data(problem_id: string): Promise<{
+  error: any | undefined
+  inputs: string[] | undefined
+  outputs: string[] | undefined
+}> {
+  const inputs: string[] = []
+  const outputs: string[] = []
+  // get the data from the database
+  await getDoc(doc(db, 'Problems', problem_id))
+    .then((problem) => {
+      if (problem.exists()) {
+        const data = problem.data().sampleData
+        for (let i = 0; i < data.length; i++) {
+          inputs.push(Buffer.from(data[i].input).toString('base64'))
+          outputs.push(Buffer.from(data[i].output).toString('base64'))
+        }
+        return { inputs: inputs, outputs: outputs, error: undefined }
+      } else {
+        return { error: 'Problem does not exist' }
+      }
+    })
+    .catch((err) => {
+      return { error: err, inputs: undefined, outputs: undefined }
+    })
+  return { inputs: inputs, outputs: outputs, error: undefined }
+}
+
 export async function submit(req: Request, res: Response) {
   const url = judge_url + '/submissions/batch?base64_encoded=true'
 
   const code = req.body.source_code
-  const inputs = req.body.inputs
-  const outputs = req.body.outputs
+  let inputs = req.body.inputs
+  let outputs = req.body.outputs
   const language_string = req.body.language_id
+  const uid = req.body.uid
+  let problem_id = req.body.problem_id
 
-  let language = 0
-  if (language_string == 'c') {
-    language = 50
-  } else if (language_string == 'cpp') {
-    language = 54
-  } else if (language_string == 'java') {
-    language = 62
-  } else if (language_string == 'py') {
-    language = 71
-  } else {
-    return res.status(400).json({ error: 'Invalid language' })
+  let error = ''
+
+  const missing: string[] = []
+  if (uid == undefined) {
+    missing.push('Missing uid')
+  }
+  if (code == undefined) {
+    missing.push('Missing source_code')
+  }
+  if (language_string == undefined) {
+    missing.push('Missing language')
   }
 
-  let compiler_flags = ''
-  let args = ''
+  if (problem_id == undefined) {
+    // arbitrary submission
 
-  // TODO: Add time/memory limits
-  // TODO: Add the submission IDs to the user's data
-  // TODO: For problems in the database, grab the data and use that for input/output
+    problem_id = -1
 
-  // C = 50, C++ = 54, Java = 62, Python: 71 (does not use pypy in default judge0)
+    if (inputs == undefined) {
+      missing.push('Missing inputs array')
+    }
 
-  if (language == 50) {
-    compiler_flags = '-g -O2 -std=c11'
-  } else if (language == 54) {
-    compiler_flags = '-g -O2 -std=c++17'
-  } else if (language == 62) {
-    args = '-Xss64m -Xmx2048m'
+    if (outputs == undefined) {
+      missing.push('Missing outputs array')
+    }
+    if (missing.length > 0) {
+      return res.status(400).json({ error: missing })
+    }
+  } else {
+    // problem submission
+    inputs = []
+    outputs = []
```

Comment:
is this necessary?

Reply:
yes, in case we are doing a problem submission, the inputs/outputs array should be undefined.